pkg/mount-manager/device-utils.go

```diff
@@ -85,75 +96,166 @@ func (m *deviceUtils) GetDiskByIdPaths(deviceName string, partition string) []st
 	}
 
-	return nil
+	return devicePath, nil
 }
 
-// Calls "udevadm trigger --action=change" on the specified drive.
-// drivePath must be the block device path to trigger on, in the format "/dev/sd*", or a symlink to it.
-// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
-func udevadmChangeToDrive(drivePath string) error {
-	klog.V(4).Infof("udevadmChangeToDrive: drive=%q", drivePath)
-
-	// Evaluate symlink, if any
-	drive, err := filepath.EvalSymlinks(drivePath)
+func udevadmTriggerForDiskIfExists(diskName string) error {
+	sds, err := filepath.Glob(diskSDPattern)
 	if err != nil {
-		return fmt.Errorf("udevadmChangeToDrive: filepath.EvalSymlinks(%q) failed with %v.", drivePath, err)
+		return fmt.Errorf("failed to filepath.Glob(\"%s\"): %v", diskSDPattern, err)
 	}
-	// Check to make sure input is "/dev/sd*"
-	if !strings.Contains(drive, diskSDPath) {
-		return fmt.Errorf("udevadmChangeToDrive: expected input in the form \"%s\" but drive is %q.", diskSDPattern, drive)
+	for _, devSDX := range sds {
+		scsiSerial, err := getScsiSerial(devSDX)
+		if err != nil {
+			return fmt.Errorf("failed to get SCSI Serial num for %s: %v", devSDX, err)
+		}
+		if scsiSerial == diskName {
+			// Found the disk that we're looking for so run a trigger on it
+			// to resolve its /dev/by-id/ path
+			klog.Warningf("udevadm --trigger running to fix disk at path %s which has SCSI ID %s", devSDX, scsiSerial)
+			err := udevadmChangeToDrive(devSDX)
+			if err != nil {
+				return fmt.Errorf("failed to fix disk at path %s which has SCSI ID %s: %v", devSDX, scsiSerial, err)
+			}
+			return nil
+		}
 	}
+	return fmt.Errorf("udevadm --trigger requested to fix disk %s but no such disk was found in %v", diskName, sds)
+}
 
+// Calls "udevadm trigger --action=change" on the specified drive. drivePath
+// must be the block device path to trigger on, in the format "/dev/sd*", or a
+// symlink to it. This is workaround for Issue #7972. Once the underlying issue
+// has been resolved, this may be removed.
+// udevadm takes a little bit to work its magic in the background so any callers
+// should not expect the trigger to complete instantly and may need to poll for
+// the change
+func udevadmChangeToDrive(devSDX string) error {
 	// Call "udevadm trigger --action=change --property-match=DEVNAME=/dev/sd..."
-	_, err = exec.Command(
+	out, err := exec.Command(
 		"udevadm",
 		"trigger",
 		"--action=change",
-		fmt.Sprintf("--property-match=DEVNAME=%s", drive)).CombinedOutput()
+		fmt.Sprintf("--property-match=DEVNAME=%s", devSDX)).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("udevadmChangeToDrive: udevadm trigger failed for drive %q with %v.", drive, err)
+		return fmt.Errorf("udevadmChangeToDrive: udevadm trigger failed for drive %q with output %s: %v.", devSDX, string(out), err)
 	}
+
+	time.Sleep(time.Second)
```

Comment:
Since this is inside a Poll loop now, do we still need this sleep?

Reply:
Not sure why this is still here. Thought I removed it, fixing